### PR TITLE
Add benchmarks for a number of NSStringAPI functions

### PR DESCRIPTION
**Pull request by phausler**
This adds a number of simple benchmarks for common NSString APIs that are exposed on String

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci please benchmark

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by moiseev** [comment]
@phausler 'please benchmark' takes forever, if you're interested in just a rough picture 'please smoke benchmark' is way faster. It will run all the benchmarks but with smaller number of samples, IIRC.

**Comment by phausler** [comment]
gotcha, I will let this run-out for now just to get good baselines.

I presume that the baselines are recorded when the tests are run? or per pull request doing an A/B execution before and after the patch?

My local executions are just hoisting the methods into a benchmark lambda:

```swift
import Foundation
#if os(macOS) || os(iOS)
import Darwin
let CLOCK_REALTIME = _CLOCK_REALTIME
#else
import Glibc
#endif

func benchmark(_ samples: Int _ work: (Int) -> Void) -> Double {
    var start = timespec()
    var end = timespec()
    clock_gettime(CLOCK_REALTIME, &start)
    work(samples)
    clock_gettime(CLOCK_REALTIME, &end)
    return (Double(end.tv_sec) * 1.0e9 + Double(end.tv_nsec)) - (Double(start.tv_sec) * 1.0e9 + Double(start.tv_nsec)) / Double(samples)
}
```

**Comment by moiseev** [comment]
I think it's the latter, as in A/B every time.

**Comment by phausler** [comment]
hmm is there a way I can cancel a build on CI?

**Comment by phausler** [comment]
@swift-ci please smoke benchmark

**Comment by moiseev** [comment]
@phausler some of the builds get cancelled automatically once the PR changes, not sure about the benchmark ones.
/cc @shahmishal can you kill the full benchmark build for this PR?

**Comment by phausler** [comment]
I was actually able to abort it

**Comment by phausler** [comment]
interesting followup results:

I was able to eek out performance improvements for the following cases:

```
run_String_capitalized
run_String_caseInsensitiveCompare_caseDiffering
run_String_caseInsensitiveCompare_equal
run_String_commonPrefix_notShared
run_String_commonPrefix_sharedPrefix_caseInsensitive
run_String_components_CharacterSet
run_String_components_String
run_String_dataUsingEncoding
run_String_enumerateLines
run_String_fastestEncoding_ascii
run_String_fastestEncoding_unicode
run_String_localizedCapitalized
```

These all were re-writes using either direct `_core` access or no-copy NSStrings created for accessing the results. Do ephemeral strings have access to `_core.startASCII` and `_core.startUTF16`?

So far I am seeing some cases be anywhere from 2x-3x up to 40x improvements. It is also worth noting that some of the tests regress when not using `_ns` (aka `as? NSString`).

**Comment by moiseev** [comment]
`_ephemeralString` is just `String`, the question though is whether obtaining the `_ephemeralString` from a `StringProtocol` will regress some of these improvements or not.

File: benchmark/single-source/NSStringAPI.swift [hunk 1]
```diff
@@ -0,0 +1,157 @@
+//===--- NSStrignAPI.swift ----------------------------------------------------===//
```
Comment (seabaylea) [inline benchmark/single-source/NSStringAPI.swift]: Same typo from #10902

**Comment by phausler** [comment]
@swift-ci please smoke benchmark

**Comment by phausler** [comment]
@swift-ci please smoke test

**Comment by phausler** [comment]
https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-smoke-perf/154/

**Comment by phausler** [comment]
We should probably add more of these to get full coverage but these should do for now to get a feeling of how NSStringAPI extensions are performing with the incoming changes

**Pull request merged**